### ISSUE TITLE
A basic protocol with "receive this many bytes" functionality

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#7577]
|<img alt="PenguinOfDoom's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| PenguinOfDoom reported|
|-|-|
|Trac ID|trac#242|
|Type|enhancement|
|Created|2003-08-27 07:08:38Z|

<details><summary>Searchable metadata</summary>

```
trac-id__242 242
type__enhancement enhancement
reporter__PenguinOfDoom PenguinOfDoom
priority__normal normal
milestone__ 
branch__ 
branch_author__ 
status__closed closed
resolution__fixed fixed
component__ 
keywords__ 
time__1061968118000000 1061968118000000
changetime__1469243308029784 1469243308029784
version__ 
owner__ 
cc__radix cc__PenguinOfDoom
```
</details>

**Comment by twisted-trac** [comment]
|<img alt="PenguinOfDoom's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|PenguinOfDoom commented|
|-|-|

```
#!html
<pre>
The state is as follows:
(next_fun, next_len)
When the buffer accumulates next_len bytes, call next_fun 
and pass it the buffer contents. The return value is the new 
state or None to drop connection.
Proposed name is twisted.protocols.basic.
ImplicitStateProtocol.
Also support stuffing bytes back onto the buffer inside 
next_fun.
For bonus points, reimplement NetstringReceiver and 
Int32Receiver in terms of this protocol.

</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="radix's avatar" src="https://avatars.githubusercontent.com/u/227068?s=50" width="50" height="50">](https://github.com/radix)|@radix commented|
|-|-|

```
#!html
<pre>
how about "LengthReceiver", or something like that? Your
proposed name doesn't make any sense to me

</pre>
```

**Comment by twisted-trac** [comment]
|<img alt="PenguinOfDoom's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|PenguinOfDoom commented|
|-|-|

```
#!html
<pre>
Well, it's in sandbox/pahan/statefulprotocol/stateful.py
The name is still crappy, because there's a t.p.basic.StatefulStringProtocol,
which does an entirely different thing.
</pre>
```

**Comment by twisted-trac** [comment]
|<img alt="PenguinOfDoom's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|PenguinOfDoom commented|
|-|-|

```
#!html
<pre>
and t.p.stateful.StatefulProtocol it is.
</pre>
```

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|<img alt="GitHub's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_6"></a>|GitHub commented|
|-|-|

In changeset 5b0977fda20855cb9070318081337c377abfb7fe
```
#!CommitTicketReference repository="" revision="5b0977fda20855cb9070318081337c377abfb7fe"
Merge pull request #242 from twisted/7868-rodrigc-rfc822

Author: rodrigc
Reviewer: markrwilliams
Fixes: #7868

Replace use of deprecated rfc822 module with email package
```
